process/smartContract/process.go

```diff
@@ -239,14 +240,22 @@ func (sc *scProcessor) processIfError(
 }
 
 func (sc *scProcessor) prepareSmartContractCall(tx data.TransactionHandler, acntSnd state.AccountHandler) error {
-	sc.isCallBack = false
+	sc.isAsyncCall = false
+	sc.isAsyncCallBack = false
 	dataToParse := tx.GetData()
 
-	scr, ok := tx.(*smartContractResult.SmartContractResult)
-	isSCRResultFromCrossShardCall := ok && len(scr.Data) > 0 && scr.Data[0] == '@'
-	if isSCRResultFromCrossShardCall {
-		dataToParse = append([]byte("callBack"), tx.GetData()...)
-		sc.isCallBack = true
+	scr, isSCR := tx.(*smartContractResult.SmartContractResult)
+	if isSCR {
+		hasTxData := scr.Data != nil && len(scr.Data) > 0
```

Comment:
len(sr.Data) > 0 is enough.  refactor these checks. Some are wrong. isScr is not needed inside.

Reply:
add a set of unit tests.

Reply:
Refactored and fixed. Also added unit test.